### PR TITLE
Changed buffer title generation logic

### PR DESCRIPTION
**Pull request by shalabhc**
- removed numeric suffix
- only file names are disambiguated
- project name is used for disambiguation if applicable
- at most one parent directory name used for disambiguation
- titles may not longer be unique (this is ok)
- the title for a buffer may automatically change when a new file
  with the same base name is opened

### CODE REVIEW

**Comment by kirbyfan64** [comment]
> - at most one parent directory name used for disambiguation

Is there a particular reason for this change? I know that, when working on any modestly-sized Python project, this will completely backfire due to the abundance of `__init__.py` files. For instance, Fbuild has `builders/c/__init__.py` and `config/c/__init__.py`, which both show up as `c/__init__.py`.

**Comment by shalabhc** [comment]
Note that the full path is also visible in the buffer list so these will not be entirely ambiguous. Also, if you had only two files open currently, they would be `__init__.py` and `c/__init__.py` which isn't any better I suppose.

While there is no absolute reason to limiting it to one parent, it followed from the following:
- Titles don't need to be unique because the full path is also shown in the list.
- I wanted titles to be consistent across all buffers with the same basename (e.g. if you have one 'parent1/foo' you will always see 'parent2/foo' and never a bare 'foo').
- It's preferable to limit the length of the title, just for aesthetics.

I assumed that one parent will be sufficient for disambiguation in almost all cases and in the remaining rare cases one can look at the full path. But I'm open to change. Do you think a hard limit of depth=2 would make more sense or do we need an unbounded prefix?

**Comment by kirbyfan64** [comment]
> Note that the full path is also visible in the buffer list so these will not be entirely ambiguous.

In my testing, they were both listed as `c/__init__.py`.

**Comment by shalabhc** [comment]
What I meant is in the second column.

File: lib/howl/buffer.moon [hunk 1]
```diff
@@ -14,22 +14,46 @@ ffi = require 'ffi'
...
-    return title if not file
-    title = file.basename .. File.separator .. title
-
-  title
-
-title_counter = (title) ->
-  title_counters[title] = 1 if not title_counters[title]
-  title_counters[title] += 1
-  title_counters[title]
+file_references = {}
```
Comment (nilnor) [inline lib/howl/buffer.moon]: I don't see how this gets cleaned up. The ref count handling does nothing to remove entries e.g. when a buffer is closed.

File: spec/buffer_spec.moon [hunk 2]
```diff
@@ -551,12 +551,8 @@ describe 'Buffer', ->
...
           assert.equal b2.title, 'sub2' .. File.separator .. 'file.foo'
 
-          sub_sub = sub1\join('sub2')
-          sub_sub\mkdir!
-          f3 = sub_sub\join('file.foo')
-          f3\touch!
-          b3.file = f3
-          assert.equal b3.title, 'sub1' .. File.separator .. b2.title
+      it 'uses the project name suffix as needed', ->
+
```
Comment (nilnor) [inline spec/buffer_spec.moon]: This is missing.

File: spec/buffer_spec.moon [hunk 3]
```diff
@@ -551,12 +551,8 @@ describe 'Buffer', ->
           b2.file = f2
           assert.equal b2.title, 'sub2' .. File.separator .. 'file.foo'
 
-          sub_sub = sub1\join('sub2')
```
Comment (nilnor) [inline spec/buffer_spec.moon]: If you remove this then the `b3` above should be removed as well.

**Comment by nilnor** [comment]
The approach taken here seems very heavy handed to me. Intuitively it feels like what should happens is that when we assign a title, we check whether this would clash with an existing title. If so, we then need to resolve the titles of the affected buffers, which would typically be a very small number of buffers. This PR resolve all the titles all the time, regardless of whether they need to be resolved or not, and it does so in a rather brute force fasion.

In short, this needs a rework. Related to this, I'm wondering whether this logic should perhaps be somewhere else, and be controlled by `Application` instead.

**Comment by shalabhc** [comment]
@nilnor - I think you have valid points. Deleting a buffer isn't handled at all and it may be best to move this logic elsewhere. I'll redo this as a new PR. I might move the logic into a new module called buffer_list so that adding and removing from the list can be tracked. There's still the problem of tracking change to the file associated with a buffer.
